### ISSUE TITLE
Follow-up: Discrepancy in the for attribute vs the id attribute of the input element

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#6505]
Follow-up to https://github.com/backdrop/backdrop-issues/issues/5968

> I see a discrepancy in the for attribute vs the id attribute of the input element. for has a value of edit-image-example-image-fid-upload vs. the input element id, which is edit-image-example-image-fid-upload--2. These two attributes need to have matching values.

**Comment by klonos** [comment]
Thank you for separating this from #5968 @herbdool 🙏🏼 

I've copied some comments from the other issue in the issue summary above, in order to provide context. I will now also move the rest of my comments (nobody else commented since)...

**Comment by klonos** [comment]
Good catch @argiepiano 👍🏼 ...if this is not achievable in php, perhaps we could use the same JS that updates the field when a file is uploaded to remove the `for` attribute (and re-add it if the file is removed).

Edit: we'd actually need to switch the `<label>` tag to a `<span>` or something, since labels w/o a matching input element is a no-no: https://stackoverflow.com/questions/33214093/is-it-good-practice-to-use-label-for-non-input-non-interactive-elements
> ...
>
> HTML specifications do not make it mandatory to associate labels with controls, but Web Content Accessibility Guidelines (WCAG) 2.0 do. This is described in the technical document [H44: Using label elements to associate text labels with form controls](https://www.w3.org/TR/WCAG20-TECHS/H44.html)...

The same article/solution suggests that `<figure>` and associated `<figcaption>` might be the most appropriate solution.

**Comment by klonos** [comment]
...for reference, in Drupal (10.1 at the time of writing this) they are using a fieldset (`<details>`) element to wrap the entire field:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/f4877f85-e79f-4eb4-8f54-8961c6affdba)

Here's the HTML while no image has been uploaded (without the wrapping `<details>` and reduced to only most of the things that matter in this case):
```html
<div class="js-form-item form-item js-form-type-managed-file form-type--managed-file js-form-item-field-image-0 form-item--field-image-0">
  <label for="edit-field-image-0-upload" id="edit-field-image-0--label" class="form-item__label">Add a new file</label>

  <div class="image-widget js-form-managed-file form-managed-file form-managed-file--image is-single has-upload no-value no-meta">
    <div class="form-managed-file__main">
      <input accept="image/*" data-drupal-selector="edit-field-image-0-upload" type="file" id="edit-field-image-0-upload" name="files[field_image_0]" size="22" class="js-form-file form-file form-element form-element--type-file form-element--api-file" data-once="fileValidate auto-file-upload"><input class="js-hide upload-button button js-form-submit form-submit" data-drupal-selector="edit-field-image-0-upload-button" formnovalidate="formnovalidate" type="submit" id="edit-field-image-0-upload-button" name="field_image_0_upload_button" value="Upload" data-once="drupal-ajax">
    </div>

    <input data-drupal-selector="edit-field-image-0-fids" type="hidden" name="field_image[0][fids]">
  </div>
  <div id="edit-field-image-0--description" class="form-item__description">
    DESCRIPTION/HELP TEXT.
  </div>
</div>
```

Here's after a file has been uploaded:
![image](https://github.com/backdrop/backdrop-issues/assets/2423362/03a60334-6932-4d54-a1b0-3b4c6c49fb79)

```html
<div class="js-form-item form-item js-form-type-managed-file form-type--managed-file js-form-item-field-image-0 form-item--field-image-0 form-item--no-label">
  <label for="edit-field-image-0-upload--UJ56gZAsCM8" id="edit-field-image-0--PK68HupcRQ0--label" class="form-item__label visually-hidden">Image</label>

  <div class="image-widget js-form-managed-file form-managed-file form-managed-file--image is-single no-upload has-value has-meta">

    <div class="form-managed-file__main">
      <span data-drupal-selector="edit-field-image-0-file-1-filename" class="file file--mime-image-jpeg file--image">
        <a href="/sites/default/files/2023-10/fingers-fighting.jpeg" type="image/jpeg">fingers-fighting.jpeg</a>
        <span class="file__size">(40.6 KB)</span>
      </span>

      <input ... REMOVE BUTTON ...>
    </div>

    <div class="form-managed-file__meta-wrapper">
      <div class="form-managed-file__meta">

        <div class="form-managed-file__image-preview image-preview">
          <div class="image-preview__img-wrapper">
            <img data-drupal-selector="edit-field-image-0-preview" src="/sites/default/files/.../fingers-fighting.jpeg?itok=VgyWWkIl" width="100" height="66" alt="" loading="lazy" class="image-style-thumbnail">
          </div>
        </div>
  
        <div class="form-managed-file__meta-items">
             ... ALT TEXT FIELD AND ITS DESCRIPTION ...
        </div>

      </div>
    </div>

    <input data-drupal-selector="edit-field-image-0-fids" type="hidden" name="field_image[0][fids]" value="1">
  </div>
</div>
```

You will notice that they have the same issue as us: there is a `<label>` with `for="edit-field-image-0-upload--UJ56gZAsCM8"`, however there is no corresponding `<input>` with that `id`. They are hiding that `<label>` element with a `visually-hidden` class.

**Comment by klonos** [comment]
...I was doing some research on this, and have found [#1734716: File field form label has incorrect id in @for attribute](https://www.drupal.org/project/drupal/issues/1734716), which then lead me to discover that most of the changes in @argiepiano's and my PR are already included with what went into D7 as part of [#1734716: [D7] Duplicate HTML IDs are created for file_managed_file fields](https://www.drupal.org/project/drupal/issues/1734716](https://www.drupal.org/project/drupal/issues/2594955)) with [this commit](https://git.drupalcode.org/project/drupal/-/commit/d07ae5d). So this makes this issue part of our D7 crossport issues as well (in the sense that won't need to backport these changes).

There's a tiny bit of that D7 commit that doesn't apply to us: the addition of `file_validate_extensions` as a JS setting via `$element['upload']['#attached']['js']` for validation purposes. We are already taking care of that via a `data-*` attribute and the changes that went into 1.4.0 core with #869 and https://github.com/backdrop/backdrop/commit/03a70011d68776632a48ea75c8ae44101dc40818.

All the above to say that we're also bring things into parity with D7 at the same time, and that the solution we are going with is what our Drupal brethren have also reviewed/tested as the most appropriate at the time. At the same time, I wasn't able to find anything in Drupal-land that solves the issue in a more ideal way.

**Comment by klonos** [comment]
@herbdool there are no conflicts in the PR, and there's a sandbox available for testing.

The solution is "half a solution", because it solves the problem of the `for` attribute pointing to the wrong element only before any file is uploaded. Once a file is uploaded, then we have the problem that @argiepiano described:
> Once you choose your file, the input element is removed, and the label therefore points at an inexistent element. ... I'm not sure what a good solution might be. There is no ideal solution here.

Well, the solution might just be to not be removing the element in the first place.

This might actually be why in Drupal they went with a wrapping fieldset/details element:
- the "Image" label goes in the entire details element, which is never removed 👍🏼 
- the "Add a new file" label is removed when the respective element it points to is removed 👍🏼

**Comment by klonos** [comment]
> ...the original `for` was pointing at the `div` rather than at the `input`.

Just to clarify: the above method of what we have been doing was wrong! ☝🏼 ...see: https://stackoverflow.com/questions/11705956/can-the-label-tags-for-attribute-be-associated-with-a-normal-div

> Some elements, not all of them form-associated, are categorized as labelable elements. These are elements that can be associated with a label element.
>
> `button` `input` (if the type attribute is not in the hidden state) `keygen` `meter` `output` `progress` `select` `textarea`

> The `for` attribute may be specified to indicate a form control with which the caption is to be associated. If the attribute is specified, the attribute's value must be the ID of a labelable element in the same Document as the label element. If the attribute is specified and there is an element in the Document whose ID is equal to the value of the for attribute, and the first such element is a labelable element, then that element is the label element's labeled control.

In short: the `for` attribute must be associated with a labelable element, and `div`s are not labelable elements.

**Comment by klonos** [comment]
...and we are already adding a fieldset when the file/image field has been configured to allow multiple/unlimited values:
<img width="614" alt="image" src="https://github.com/backdrop/backdrop-issues/assets/2423362/a4facb88-4515-49c7-bd16-09efcdee2a32">

We might as well be adding a fieldset even if a single file is allowed. That would solve this.
